File: behavioral-design-pattern/template-pattern/template.ts
/**
 * Template Method Pattern
 * ========================
 * Category: Behavioral Design Pattern
 *
 * Intent:
 *   Define the skeleton of an algorithm in a base class, deferring some steps
 *   to subclasses. Template Method lets subclasses redefine certain steps of
 *   an algorithm without changing the algorithm's structure.
 *
 * When to use:
 *   - You want to implement the invariant parts of an algorithm once and let
 *     subclasses provide the behavior that varies
 *   - Common behavior among subclasses should be factored and localized in a
 *     single class to avoid duplication
 *   - You want to control which steps subclasses can extend (hooks vs. required)
 *
 * Key Participants:
 *   - AbstractClass (DataMiner): defines the template method and abstract steps
 *   - ConcreteClass (CSVMiner, JSONMiner, DatabaseMiner): implements the steps
 *   - Template Method (mine): the algorithm skeleton that calls the steps in order
 *
 * Hollywood Principle:
 *   "Don't call us, we'll call you." -- The base class calls the subclass methods,
 *   not the other way around. Subclasses plug into the algorithm; they don't drive it.
 */

// --- Abstract Class ---

/**
 * DataMiner defines the template method `mine()`.
 * The algorithm skeleton is fixed; subclasses override individual steps.
 */
abstract class DataMiner {
  /**
   * Template method: defines the algorithm skeleton.
   * This method should NOT be overridden by subclasses.
   */
  mine(source: string): void {
    console.log(`\n${"=".repeat(60)}`);
    console.log(`  Mining data from: ${source}`);
    console.log("=".repeat(60));

    const rawData = this.readData(source);
    const parsed = this.parseData(rawData);

    // Hook: optional pre-analysis step (subclasses may override)
    this.beforeAnalysis(parsed);

    const results = this.analyzeData(parsed);
    this.generateReport(results);

    // Hook: optional cleanup step
    this.cleanup();

    console.log("  Pipeline complete.\n");
  }

  // --- Abstract steps (subclasses MUST implement) ---

  /** Read raw data from the source. */
  protected abstract readData(source: string): string;

  /** Parse raw data into structured records. */
  protected abstract parseData(rawData: string): Record<string, string | number>[];

  /** Analyze parsed data and produce results. */
  protected abstract analyzeData(data: Record<string, string | number>[]): Record<string, unknown>;

  /** Generate a report from the analysis results. */
  protected abstract generateReport(results: Record<string, unknown>): void;

  // --- Hook methods (subclasses CAN override, but don't have to) ---

  /** Hook: called before analysis. Override to add pre-processing. */
  protected beforeAnalysis(_data: Record<string, string | number>[]): void {
    // Default: do nothing
  }

  /** Hook: called after the pipeline. Override to release resources. */
  protected cleanup(): void {
    // Default: do nothing
  }
}

// --- Concrete Classes ---

/** Mines data from CSV files. */
class CSVMiner extends DataMiner {
  protected readData(source: string): string {
    console.log(`  [CSV] Reading file: ${source}`);
    // Simulated CSV content
    return "name,age,city\nAlice,30,Paris\nBob,25,London\nCharlie,35,Berlin";
  }

  protected parseData(rawData: string): Record<string, string>[] {
    console.log("  [CSV] Parsing CSV rows...");
    const lines = rawData.trim().split("\n");
    const headers = lines[0].split(",");
    const records = lines.slice(1).map((line) => {
      const values = line.split(",");
      const record: Record<string, string> = {};
      headers.forEach((h, i) => (record[h] = values[i]));
      return record;
    });
    console.log(`  [CSV] Parsed ${records.length} records`);
    return records;
  }

  protected analyzeData(data: Record<string, string>[]): Record<string, number> {
    console.log("  [CSV] Analyzing age distribution...");
    const ages = data.map((r) => parseInt(r.age, 10));
    return {
      count: ages.length,
      average_age: ages.reduce((a, b) => a + b, 0) / ages.length,
      min_age: Math.min(...ages),
      max_age: Math.max(...ages),
    };
  }

  protected generateReport(results: Record<string, number>): void {
    console.log("  [CSV] Report:");
    for (const [key, value] of Object.entries(results)) {
      console.log(`    ${key}: ${value}`);
    }
  }
}

/** Mines data from JSON sources. */
class JSONMiner extends DataMiner {
  protected readData(source: string): string {
    console.log(`  [JSON] Fetching JSON from: ${source}`);
    return '[{"product":"Widget","price":9.99},{"product":"Gadget","price":24.99},{"product":"Doohickey","price":4.99}]';
  }

  protected parseData(rawData: string): Record<string, string | number>[] {
    console.log("  [JSON] Parsing JSON array...");
    const records = JSON.parse(rawData) as Record<string, string | number>[];
    console.log(`  [JSON] Parsed ${records.length} records`);
    return records;
  }

  protected analyzeData(data: Record<string, string | number>[]): Record<string, number> {
    console.log("  [JSON] Analyzing price data...");
    const prices = data.map((r) => r.price as number);
    return {
      total_products: prices.length,
      total_revenue_potential: prices.reduce((a, b) => a + b, 0),
      cheapest: Math.min(...prices),
      most_expensive: Math.max(...prices),
    };
  }

  protected generateReport(results: Record<string, number>): void {
    console.log("  [JSON] Report:");
    for (const [key, value] of Object.entries(results)) {
      console.log(`    ${key}: ${value}`);
    }
  }

  // Override hook: validate JSON structure before analysis
  protected beforeAnalysis(data: Record<string, string | number>[]): void {
    console.log("  [JSON] Validating schema: all records have 'product' and 'price'...");
    for (const record of data) {
      if (!("product" in record) || !("price" in record)) {
        throw new Error(`Invalid record: ${JSON.stringify(record)}`);
      }
    }
    console.log("  [JSON] Schema validation passed.");
  }
}

/** Mines data from a database connection. */
class DatabaseMiner extends DataMiner {
  private connectionOpen = false;

  protected readData(source: string): string {
    console.log(`  [DB] Connecting to database: ${source}`);
    this.connectionOpen = true;
    console.log("  [DB] Executing query: SELECT * FROM users");
    return "id:1,name:Alice,role:admin|id:2,name:Bob,role:user|id:3,name:Charlie,role:user";
  }

  protected parseData(rawData: string): Record<string, string>[] {
    console.log("  [DB] Parsing query results...");
    const records = rawData.split("|").map((row) => {
      const record: Record<string, string> = {};
      row.split(",").forEach((field) => {
        const [key, value] = field.split(":");
        record[key] = value;
      });
      return record;
    });
    console.log(`  [DB] Parsed ${records.length} records`);
    return records;
  }

  protected analyzeData(data: Record<string, string>[]): Record<string, unknown> {
    console.log("  [DB] Analyzing user roles...");
    const roles: Record<string, number> = {};
    for (const record of data) {
      roles[record.role] = (roles[record.role] || 0) + 1;
    }
    return { total_users: data.length, role_distribution: roles };
  }

  protected generateReport(results: Record<string, unknown>): void {
    console.log("  [DB] Report:");
    console.log(`    Total users: ${results.total_users}`);
    const dist = results.role_distribution as Record<string, number>;
    for (const [role, count] of Object.entries(dist)) {
      console.log(`    ${role}: ${count}`);
    }
  }

  // Override hook: close the database connection
  protected cleanup(): void {
    if (this.connectionOpen) {
      console.log("  [DB] Closing database connection...");
      this.connectionOpen = false;
    }
  }
}

// --- Main ---

function main(): void {
  // Each miner follows the same algorithm (mine), but the steps differ.
  // The base class controls the flow -- Hollywood Principle in action.

  const pipelines: [DataMiner, string][] = [
    [new CSVMiner(), "users.csv"],
    [new JSONMiner(), "https://api.example.com/products"],
    [new DatabaseMiner(), "postgres://localhost:5432/mydb"],
  ];

  for (const [miner, source] of pipelines) {
    miner.mine(source);
  }
}

main();

File: behavioral-design-pattern/template-pattern/template.py
"""
Template Method Pattern
========================
Category: Behavioral Design Pattern

Intent:
    Define the skeleton of an algorithm in a base class, deferring some steps
    to subclasses. Template Method lets subclasses redefine certain steps of
    an algorithm without changing the algorithm's structure.

When to use:
    - You want to implement the invariant parts of an algorithm once and let
      subclasses provide the behavior that varies
    - Common behavior among subclasses should be factored and localized in a
      single class to avoid duplication
    - You want to control which steps subclasses can extend (hooks vs. required)

Key Participants:
    - AbstractClass (DataMiner): defines the template method and abstract steps
    - ConcreteClass (CSVMiner, JSONMiner, DatabaseMiner): implements the steps
    - Template Method (mine): the algorithm skeleton that calls the steps in order

Hollywood Principle:
    "Don't call us, we'll call you." -- The base class calls the subclass methods,
    not the other way around. Subclasses plug into the algorithm; they don't drive it.
"""

from __future__ import annotations
from abc import ABC, abstractmethod


# --- Abstract Class ---

class DataMiner(ABC):
    """
    Abstract class that defines the template method `mine()`.
    The algorithm skeleton is fixed; subclasses override individual steps.
    """

    def mine(self, source: str) -> None:
        """
        Template method: defines the algorithm skeleton.
        This method is final -- subclasses should NOT override it.
        """
        print(f"\n{'='*60}")
        print(f"  Mining data from: {source}")
        print(f"{'='*60}")

        raw_data = self.read_data(source)
        parsed = self.parse_data(raw_data)

        # Hook: optional pre-analysis step (subclasses may override)
        self.before_analysis(parsed)

        results = self.analyze_data(parsed)
        self.generate_report(results)

        # Hook: optional cleanup step
        self.cleanup()

        print("  Pipeline complete.\n")

    # --- Abstract steps (subclasses MUST implement) ---

    @abstractmethod
    def read_data(self, source: str) -> str:
        """Read raw data from the source."""
        pass

    @abstractmethod
    def parse_data(self, raw_data: str) -> list[dict]:
        """Parse raw data into structured records."""
        pass

    @abstractmethod
    def analyze_data(self, data: list[dict]) -> dict:
        """Analyze parsed data and produce results."""
        pass

    @abstractmethod
    def generate_report(self, results: dict) -> None:
        """Generate a report from the analysis results."""
        pass

    # --- Hook methods (subclasses CAN override, but don't have to) ---

    def before_analysis(self, data: list[dict]) -> None:
        """Hook: called before analysis. Override to add pre-processing."""
        pass

    def cleanup(self) -> None:
        """Hook: called after the pipeline. Override to release resources."""
        pass


# --- Concrete Classes ---

class CSVMiner(DataMiner):
    """Mines data from CSV files."""

    def read_data(self, source: str) -> str:
        print(f"  [CSV] Reading file: {source}")
        # Simulated CSV content
        return "name,age,city\nAlice,30,Paris\nBob,25,London\nCharlie,35,Berlin"

    def parse_data(self, raw_data: str) -> list[dict]:
        print("  [CSV] Parsing CSV rows...")
        lines = raw_data.strip().split("\n")
        headers = lines[0].split(",")
        records = []
        for line in lines[1:]:
            values = line.split(",")
            records.append(dict(zip(headers, values)))
        print(f"  [CSV] Parsed {len(records)} records")
        return records

    def analyze_data(self, data: list[dict]) -> dict:
        print("  [CSV] Analyzing age distribution...")
        ages = [int(r["age"]) for r in data]
        return {
            "count": len(ages),
            "average_age": sum(ages) / len(ages),
            "min_age": min(ages),
            "max_age": max(ages),
        }

    def generate_report(self, results: dict) -> None:
        print("  [CSV] Report:")
        for key, value in results.items():
            print(f"    {key}: {value}")


class JSONMiner(DataMiner):
    """Mines data from JSON sources."""

    def read_data(self, source: str) -> str:
        print(f"  [JSON] Fetching JSON from: {source}")
        # Simulated JSON content
        return '[{"product":"Widget","price":9.99},{"product":"Gadget","price":24.99},{"product":"Doohickey","price":4.99}]'

    def parse_data(self, raw_data: str) -> list[dict]:
        import json
        print("  [JSON] Parsing JSON array...")
        records = json.loads(raw_data)
        print(f"  [JSON] Parsed {len(records)} records")
        return records

    def analyze_data(self, data: list[dict]) -> dict:
        print("  [JSON] Analyzing price data...")
        prices = [r["price"] for r in data]
        return {
            "total_products": len(prices),
            "total_revenue_potential": sum(prices),
            "cheapest": min(prices),
            "most_expensive": max(prices),
        }

    def generate_report(self, results: dict) -> None:
        print("  [JSON] Report:")
        for key, value in results.items():
            print(f"    {key}: {value}")

    # Override hook: validate JSON structure before analysis
    def before_analysis(self, data: list[dict]) -> None:
        print("  [JSON] Validating schema: all records have 'product' and 'price'...")
        for record in data:
            assert "product" in record and "price" in record, f"Invalid record: {record}"
        print("  [JSON] Schema validation passed.")


class DatabaseMiner(DataMiner):
    """Mines data from a database connection."""

    def __init__(self) -> None:
        self._connection_open = False

    def read_data(self, source: str) -> str:
        print(f"  [DB] Connecting to database: {source}")
        self._connection_open = True
        print("  [DB] Executing query: SELECT * FROM users")
        # Simulated query result as string
        return "id:1,name:Alice,role:admin|id:2,name:Bob,role:user|id:3,name:Charlie,role:user"

    def parse_data(self, raw_data: str) -> list[dict]:
        print("  [DB] Parsing query results...")
        records = []
        for row in raw_data.split("|"):
            record = {}
            for field in row.split(","):
                key, value = field.split(":")
                record[key] = value
            records.append(record)
        print(f"  [DB] Parsed {len(records)} records")
        return records

    def analyze_data(self, data: list[dict]) -> dict:
        print("  [DB] Analyzing user roles...")
        roles: dict[str, int] = {}
        for record in data:
            role = record["role"]
            roles[role] = roles.get(role, 0) + 1
        return {"total_users": len(data), "role_distribution": roles}

    def generate_report(self, results: dict) -> None:
        print("  [DB] Report:")
        print(f"    Total users: {results['total_users']}")
        for role, count in results["role_distribution"].items():
            print(f"    {role}: {count}")

    # Override hook: close the database connection
    def cleanup(self) -> None:
        if self._connection_open:
            print("  [DB] Closing database connection...")
            self._connection_open = False


# --- Main ---

def main() -> None:
    # Each miner follows the same algorithm (mine), but the steps differ.
    # The base class controls the flow -- Hollywood Principle in action.

    miners: list[tuple[DataMiner, str]] = [
        (CSVMiner(), "users.csv"),
        (JSONMiner(), "https://api.example.com/products"),
        (DatabaseMiner(), "postgres://localhost:5432/mydb"),
    ]

    for miner, source in miners:
        miner.mine(source)


if __name__ == "__main__":
    main()
